Make request transport for amp-analytics configurable.

btownsend opened the pull request:
Removes `host` config attribute.

Adds new `transport` config attribute whose value is an array of request transport mechanisms, in order of preference. The first transport in the list available on the user's browser will be used. The default `transport` value is set to `["beacon", "xhrpost", "image"]` but can be overridden in the built-in vendor config objects and/or remote config and/or inline config.

Alternative to the implementation in #1162

dvoytenko commented on extensions/amp-analytics/0.1/transport.js:
```diff
@@ -0,0 +1,105 @@
+  if (!win.navigator.sendBeacon) {
+    return false;
+  }
+  win.navigator.sendBeacon(request, '');
```
I believe sendBeacon returns true/false based on whether it succeeds. Should we use it to continue the fallback?

btownsend replied:
Done. A false return value means the queue has been exhausted. Not sure whether it's better to throttle (by dropping the request) or fallback. I'm going with just falling back for now. We should probably revisit this as part of a more comprehensive look at request throttling.

dvoytenko commented on extensions/amp-analytics/0.1/transport.js:
```diff
@@ -0,0 +1,105 @@
+  if (!win.XMLHttpRequest) {
+    return false;
+  }
+  const xhr = new win.XMLHttpRequest();
```
This is pretty incomplete. We need to find a way to use our `xhr.js` - it has a bunch of polyfills for `fetch`.

btownsend replied:
I had wanted to build off xhr.js originally but it looked like it would be kind of a mess to support. For this case, we don't want the XDomainObject fallback (since that will not get us cookies.) Our success criteria is also just whether the browser supports XHR+CORS (rather than knowing the outcome of the request).

If you feel strongly about extending xhr.js to support these requirements, or have any ideas on the best way to approach that, please let me know!

dvoytenko replied:
Why wouldn't XDomainObject get us cookies? Isn't its main purpose?

`xhr.js` main purpose is to polyfill Fetch API which, if supported by the browser, would cover all the cases that we need. We just need to make sure that our polyfill does the right thing, and I think, it's either already there or very close. Currently the `fetch_` method is private since we didn't have a need before to make it public, but we can export it here as needed.

Let's try to do so and see what's our delta. One thing though - if you are using Chrome for testing, it's kind of hard to test since Chrome implements Fetch API natively and so the things would likely appear great immediately but can later fail on some other browser. So, we just need to make sure we test polyfill'd case sufficiently.

btownsend replied:
No, XDomainRequest does not include cookies - it was for cross-domain XHR. It has a few other restrictions that can be a challenge too (e.g. the request protocol must match the page protocol.) This page is a good resource: http://blogs.msdn.com/b/ieinternals/archive/2010/05/13/xdomainrequest-restrictions-limitations-and-workarounds.aspx

If we can remove XDomainRequest completely from xhr.js that would make things easier, but would limit support to IE 10+. Is that acceptable?

dvoytenko replied:
You're right - looks like XDomainRequest. Alright. This means we will have to live with this then.

dvoytenko commented on extensions/amp-analytics/0.1/vendors.js:
```diff
@@ -46,17 +47,20 @@ export const ANALYTICS_CONFIG = {
-          'el=${eventLabel}&ev=${eventValue}',
-      'social': '${baseHit}&t=social&sa=${socialAction}&sn=${socialNetwork}&' +
-          'st=${socialActionTarget}'
+      'commonHost': 'https://www.google-analytics.com',
```
Why "common" in names?

btownsend replied:
Those are request parts common to all the other requests. (The expansion scope of these are just the requests object itself.)

dvoytenko replied:
Ok. But why did you decide to get rid of the "host"? I'm a bit behind on some of the thinking here, but my rule of thumb, "If you want to call something 'common', first call it 'trash' and if it still works - go ahead" :)

btownsend replied:
I removed `host` just as a simplification - it seemed arbitrary that `host` was a special top level value while the path+params+etc was a separate string. Combining them into a URL seemed simpler while also providing some extra flexibility since they can now be expanded by placeholders like the rest of the string.

I could rename `commonHost` to `host` above if that's clearer. The behavior will be the same.

dvoytenko replied:
I think renaming to `host` would be better.

dvoytenko commented:
Ok. I think we are very close then. Just some naming to be confirmed.

btownsend commented:
Updated based on earlier feedback/chats.

The `transport` config attribute now takes an object value that can individually specify if `beacon`, `xhrpost`, and `image` transport is allowed. The precedence is fixed as beacon > xhrpost > image.

Added tests.

Ready for another review.

dvoytenko commented on examples/analytics.amp.html:
```diff
@@ -22,9 +22,9 @@
 <script type="application/json">
 {
-  "host": "example.com",
+  "transport": {"beacon": false, "xhrpost": false},
```
Is the intention that "image" is enabled by default? From what I can tell in the code, this is not the case...

btownsend replied:
In vendors.js, the "default" config (which all configs inherit from) has `beacon`, `xhrpost`, and `image` all set to `true`. This means the default behavior will be beacon > xhrpost > image unless one of those options are specifically disabled in the vendor, inline, or remote config.

dvoytenko commented:
Looks good. One comment I made above about transport defaults. Also, are the docs changing as well?

btownsend commented:
For the docs I'm planning to do a full pass tomorrow morning in a separate PR to include this change, remote config change (#1174 needs to be merged please!), the trigger change (already merged), and probably some other things I've missed.

dvoytenko commented:
Ok. Merging this PR and looking forward to docs update.

dvoytenko commented on extensions/amp-analytics/0.1/transport.js:
```diff
@@ -0,0 +1,110 @@
+   */
+  static sendRequestUsingImage(win, request) {
+    const image = new Image();
+    image.src = request;
```
@btownsend do we automatically add random cache buster to URLs? Or is it up to the publisher?

kzap replied:
the URL would be up to the publisher configs, best not to make much assumptions but give flexibility then provide assumptions in default configs in `vendors.js`

kzap commented:
I really like the discussion in #1162 to add support for POST data
There are a number of analytics providers that look for POST data in the body whether its a json payload or actual form encoded data in the body.
